lib/api/model/submessage.dart

```diff
@@ -0,0 +1,285 @@
+    '${senderId ?? 'canned'},$optionIndex';
+
+  factory PollEvent.fromJson(Map<String, Object?> json) {
+    final rawPollEventType = json['type'];
+    if (rawPollEventType == null) return UnknownPollEvent(json: json);
+    switch ($enumDecode(
+      _$PollEventTypeEnumMap,
+      rawPollEventType,
+      unknownValue: PollEventType.unknown,
+    )) {
+      case PollEventType.newOption: return PollOptionEvent.fromJson(json);
+      case PollEventType.question: return PollQuestionEvent.fromJson(json);
+      case PollEventType.vote: return PollVoteEvent.fromJson(json);
+      case PollEventType.unknown: return UnknownPollEvent(json: json);
+    }
+  }
+
+  Map<String, Object?> toJson();
+}
+
+/// A poll event when an option is added.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollOptionEvent extends PollEvent {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventType get type => PollEventType.newOption;
+
+  final String option;
+  /// The index of last [option] added by the sender.
+  @JsonKey(name: 'idx')
+  final int latestOptionIndex;
+
+  PollOptionEvent({required this.option, required this.latestOptionIndex});
+
+  @override
+  factory PollOptionEvent.fromJson(Map<String, Object?> json) =>
+    _$PollOptionEventFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollOptionEventToJson(this);
+}
+
+/// A poll event when the question has been edited.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollQuestionEvent extends PollEvent {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventType get type => PollEventType.question;
+
+  final String question;
+
+  PollQuestionEvent({required this.question});
+
+  @override
+  factory PollQuestionEvent.fromJson(Map<String, Object?> json) =>
+    _$PollQuestionEventFromJson(json);
+
+  @override
+  Map<String, Object?> toJson() => _$PollQuestionEventToJson(this);
+}
+
+/// A poll event when a vote has been cast or removed.
+@JsonSerializable(fieldRename: FieldRename.snake)
+class PollVoteEvent extends PollEvent {
+  @override
+  @JsonKey(includeToJson: true)
+  PollEventType get type => PollEventType.vote;
+
+  /// The key of the affected option.
+  ///
+  /// See [PollEvent.optionKey].
+  final String key;
+  @JsonKey(name: 'vote', unknownEnumValue: VoteOp.unknown)
+  final VoteOp op;
+
+  PollVoteEvent({required this.key, required this.op});
+
+  @override
+  factory PollVoteEvent.fromJson(Map<String, Object?> json) {
+    final result = _$PollVoteEventFromJson(json);
+    // Crunchy-shell validation
+    final segments = result.key.split(',');
+    final [senderId, optionIndex] = segments;
+    if (senderId != 'canned') {
+      int.parse(senderId, radix: 10);
+    }
+    int.parse(optionIndex, radix: 10);
+    return result;
+  }
+
+  @override
+  Map<String, Object?> toJson() => _$PollVoteEventToJson(this);
+}
+
+/// As in [PollVoteEvent.op].
+@JsonEnum(valueField: 'apiValue')
+enum VoteOp {
```

Comment:
Similar to the comment on `Option`:
```suggestion
enum PollVoteOp {
```